src/io/flutter/utils/AndroidUtils.java

```diff
@@ -114,4 +185,451 @@ private static String validateName(String name, boolean requiresSeparator) {
+        modules.forEach((module) -> {
+          if (!DartSdkLibUtil.isDartSdkEnabled(module)) {
+            new EnableDartSupportForModule(module).run();
+          }
+        });
+      }, 1, TimeUnit.SECONDS);
+    });
+  }
+
+  private static boolean isVanillaAddToApp(@NotNull Project project, @Nullable VirtualFile file, @NotNull String name) {
+    if (file == null) {
+      return false;
+    }
+    VirtualFile dir = file.getParent();
+    if (dir.findChild(".ios") == null) {
+      dir = dir.getParent();
+    }
+    if (dir.getName().equals(".android")) {
+      dir = dir.getParent();
+    }
+    if (dir.findChild(".ios") == null && dir.findChild("ios") == null) {
+      return false;
+    }
+    if (doesBuildFileExist(dir)) {
+      return true;
+    }
+    GradleSettingsFile parsedSettings = parseSettings(project);
+    // Check settings for "include :name".
+    return findInclude(requireNonNull(parsedSettings), name) == null;
+  }
+
+  private static boolean doesBuildFileExist(VirtualFile dir) {
+    return new File(dir.getPath(), SdkConstants.FN_BUILD_GRADLE).exists();
+  }
+
+  private static GradleSettingsFile parseSettings(Project project) {
+    // Get the PSI for settings.gradle. In IntelliJ it is just this, but Android Studio uses a VirtualFile.
+    //GradleSettingsFile parsedSettings =
+    //  BuildModelContext.create(project).getOrCreateSettingsFile(project);
+    boolean isAndroidStudio = FlutterUtils.isAndroidStudio();
+    VirtualFile projectDir = requireNonNull(FlutterUtils.getProjectRoot(project));
+    Object param = isAndroidStudio ? projectDir.findChild(SdkConstants.FN_SETTINGS_GRADLE) : project;
+    if (param == null) {
+      return null;
+    }
+    Method method =
+      ReflectionUtil.getMethod(BuildModelContext.class, "getOrCreateSettingsFile", isAndroidStudio ? VirtualFile.class : Project.class);
+    if (method == null) {
+      return null;
+    }
+    try {
+      return (GradleSettingsFile)method.invoke(BuildModelContext.create(project), param);
+    }
+    catch (InvocationTargetException | IllegalAccessException e) {
+      return null;
+    }
+  }
+
+  // The project is an Android project that contains a Flutter module.
+  private static void enableCoEditing(@NotNull Project project) {
+    Module module = FlutterUtils.findFlutterGradleModule(project);
+    if (module == null) return;
+    VirtualFile moduleFile = module.getModuleFile();
+    if (moduleFile == null) return;
+    VirtualFile androidDir = moduleFile.getParent().getParent();
+    VirtualFile flutterModuleDir = androidDir.getParent();
+    String flutterModuleName = flutterModuleDir.getName();
+    if (!isVanillaAddToApp(project, androidDir, flutterModuleName)) return;
+    new CoEditHelper(project, flutterModuleDir).enable();
+  }
+
+  private static void addCoeditTransformedProject(@NotNull Project project) {
+    if (!project.isDisposed()) {
+      COEDIT_TRANSFORMED_PROJECTS.add(project);
```

Comment:
Correct, the weak reference allows a disposed project to get GC'd, and the garbage collector removes it from the collection. This will be an infrequent occurrence.